src/mqtt_client.rs

```diff
@@ -365,27 +390,96 @@ where
                 Ok(())
             }
 
+            ReceivedPacket::PingResp => {
+                self.state.borrow_mut().pending_pingreq_time.take();
+                Ok(())
+            }
+
             _ => Err(Error::Unsupported),
         }
     }
 
+    fn open_socket(&mut self) -> Result<(), N::Error> {
+        let socket = self.network_stack.open(Mode::NonBlocking)?;
+        self.socket.borrow_mut().replace(socket);
+        Ok(())
+    }
+
     fn connect_socket(&mut self) -> Result<(), Error<N::Error>> {
+        // Don't try connecting more than once per second to avoid network spam if
+        // the broker port is closed.
+        if let Some(last_connect) = &self.last_socket_connect_time {
+            let now = self.clock.try_now().unwrap();
+            if now
+                .checked_duration_since(last_connect)
+                .and_then(|diff| Seconds::<u32>::try_from(diff).ok())
```

Comment:
Add a note as to why calling `ok()` on this is acceptable - this would fail and result in weird behavior if the diff were negative.

Reply:
(I will copy/paste the same comment as for the keepalive logic, though.)